Libraries/CommonFunctions.psm1

```diff
@@ -150,30 +120,37 @@ Function Select-TestCases($TestXMLs, $TestCategory, $TestArea, $TestNames, $Test
+            # if TestCategory not provided, or test case has Category completely matching one of expected TestCategory (case insensitive 'notcontains'), otherwise continue (skip this test case)
             if (($TestCategory -ne "*") -and ($testCategoryArray -notcontains $test.Category)) {
                 continue
             }
-            # case insensitive 'notcontains', and only select complete matched with the concatenated expression for Area of TestCase, otherwise skip (continue)
+            # if TestArea not provided, or test case has Area completely matching one of expected TestArea (case insensitive 'notcontains'), otherwise continue (skip this test case)
             if (($TestArea -ne "*") -and ($testAreaArray -notcontains $test.Area)) {
                 continue
             }
-            # only select TestCase that has SetupType value insensitive matching the expected Setup pattern, otherwise skip (continue)
-            if (($TestSetup -ne "*") -and !$($testSetupTypeArray | Where-Object {"$($test.SetupConfig.SetupType)" -imatch $_})) {
+            # if TestSetup not provided, or test case has SetupType value defined which could insensitively match one of expected Setup pattern, otherwise continue (skip this test case)
+            if (($TestSetup -ne "*") -and !$($testSetupTypeArray | Where-Object {$test.SetupConfig.SetupType -and ("$($test.SetupConfig.SetupType)" -imatch $_)})) {
                 continue
             }
-            # case insensitive 'notcontains', and only select complete matched with the concatenated expression for TestName of TestCase, otherwise skip (continue)
+            # if TestNames not provided, or test case has TestName completely matching one of expected TestNames (case insensitive 'notcontains'), otherwise continue (skip this test case)
             if (($TestNames -ne "*") -and ($testNamesArray -notcontains $test.testName)) {
                 continue
             }
 
-            $testTagMatched = Match-TestTag -currentTest $test -TestTag $TestTag
-            if ($testTagMatched -eq $false) {
-                continue
+            # if TestTag not provided, or test case has Tags not defined (backward compatible), or test case has defined Tags scope and this Tags scope has value completely matching one of expected TestTag (case insensitive 'contains'), otherwise continue (skip this test case)
+            if (($TestTag -ne "*")) {
+                if (!$test.Tags) {
+                    Write-LogWarn "Test Tags of $($test.TestName) is not defined; include this test case by default."
+                }
+                elseif (!$($testTagArray | Where-Object {$test.Tags.Trim(', ').Split(',').Trim() -contains $_})) {
+                    continue
+                }
             }
-
-            $testPriorityMatched = Match-TestPriority -currentTest $test -TestPriority $TestPriority
```

Comment:
Any bug prompted this change ?

Reply:
align the code in the central function (previous separated functions also only used here), and no need to duplicate splitting TestPriority and TestTag string for every test case from every test definition xml.